osbenchmark/telemetry.py

```diff
@@ -1669,3 +1670,149 @@ def detach_from_node(self, node, running):
+            recorder = SegmentReplicationStatsRecorder(
+                cluster_name, self.clients[cluster_name], self.metrics_store, self.sample_interval,
+                self.indices_per_cluster[cluster_name] if self.indices_per_cluster else None)
+            sampler = SamplerThread(recorder)
+            self.samplers.append(sampler)
+            sampler.setDaemon(True)
+            # we don't require starting recorders precisely at the same time
+            sampler.start()
+
+    def on_benchmark_stop(self):
+        if self.samplers:
+            for sampler in self.samplers:
+                sampler.finish()
+
+class SegmentReplicationStatsRecorder:
+    """
+    Collects and pushes segment replication stats for the specified cluster to the metric store.
+    """
+
+    def __init__(self, cluster_name, client, metrics_store, sample_interval, indices=None):
+        """
+        :param cluster_name: The cluster_name that the client connects to, as specified in target.hosts.
+        :param client: The OpenSearch client for this cluster.
+        :param metrics_store: The configured metrics store we write to.
+        :param sample_interval: integer controlling the interval, in seconds, between collecting samples.
+        :param indices: optional list of indices to filter results from.
+        """
+
+        self.cluster_name = cluster_name
+        self.client = client
+        self.metrics_store = metrics_store
+        self.sample_interval = sample_interval
+        self.indices = indices
+        self.logger = logging.getLogger(__name__)
+
+    def __str__(self):
+        return "segment replication stats"
+
+    def record(self):
+        """
+        Collect segment replication stats for indexes (optionally) specified in telemetry parameters
+        and push to metrics store.
+        """
+        if not self.indices:
+            # if 'index' parameter is not assigned, set it as empty string to allow getting API response for all indexes
+            self.indices = [""]
+        for index in self.indices:
+            try:
+                stats_api_endpoint = "/_cat/segment_replication/"
+                stats = self.client.transport.perform_request("GET", stats_api_endpoint + index, params={"time": "ms"})
+            except opensearchpy.TransportError:
+                raise exceptions.BenchmarkError(
+                    f"A transport error occurred while collecting segment replication stats on cluster "
+                    f"[{self.cluster_name}]") from None
+
+            # parse the REST API response, each field will be an array element
+            stats_arr = []
+            for line_of_shard_stats in stats.splitlines():
+                stats_arr.append(line_of_shard_stats.split(" "))
+
+            for shard_stats_arr in stats_arr:
+                self._push_stats(stats=shard_stats_arr, index=index)
+
+    def _push_stats(self, stats, index=None):
+
+        doc = {
+            "name": "segment-replication-stats",
```

Comment:
In addition to lag,can we also include `target_node` so we can filter by a specific shard & node,  `bytes_behind` and `rejected_requests` ?

Reply:
@mch2  Thanks for your review! I got it, since those fields are all included in the REST API response, it will be simple to add them all into the metrics output.
Seeing from the API document https://opensearch.org/docs/latest/api-reference/cat/cat-segment-replication/, there is a "shards" parameter available to filter the shards to be shown in the API response, I can also add the parameter to filter the metrics result.

Reply:
I added all the fields from the REST API response in a new commit (https://github.com/opensearch-project/opensearch-benchmark/pull/346/commits/281c2c4a51aaaa2ac840eaade8633e23c947751e)